File: Tutorials/BiometricStandards/Java/cbeff-record-to-ntemplate/src/main/java/com/neurotec/tutorials/CBEFFRecordToNTemplate.java
package com.neurotec.tutorials;

import java.io.IOException;

import com.neurotec.biometrics.NBiometricEngine;
import com.neurotec.biometrics.NBiometricStatus;
import com.neurotec.biometrics.NSubject;
import com.neurotec.biometrics.standards.CBEFFRecord;
import com.neurotec.io.NBuffer;
import com.neurotec.io.NFile;
import com.neurotec.licensing.NLicense;
import com.neurotec.licensing.NLicenseManager;
import com.neurotec.tutorials.util.LibraryManager;
import com.neurotec.tutorials.util.Utils;

public class CBEFFRecordToNTemplate {

	private static final String DESCRIPTION = "Converting CbeffRecord to NTemplate";
	private static final String NAME = "complex-cbeff-record";
	private static final String VERSION = "13.1.0.0";

	private static void usage() {
		System.out.println("usage:");
		System.out.format("\t%s [CbeffRecord] [PatronFormat] [NTemplate] ...", NAME);
		System.out.println();
		System.out.println("\t[CbeffRecord] - filename of CbeffRecord.");
		System.out.println("\t[PatronFormat] - number identifying patron format (all supported values can be found in CbeffRecord class documentation).");
		System.out.println("\t[NTemplate] - filename of NTemplate.");
		System.out.println();
	}

	public static void main(String[] args) {
		LibraryManager.initLibraryPath();
		Utils.printTutorialHeader(DESCRIPTION, NAME, VERSION, args);

		if (args.length != 3) {
			usage();
			System.exit(1);
		}

		//=========================================================================
		// CHOOSE LICENCES !!!
		//=========================================================================
		// ONE of the below listed "licenses" lines is required for unlocking this sample's functionality. Choose licenses that you currently have on your device.
		// If you are using a TRIAL version - choose any of them.

		final String[] licenses = { "FingerClient", "PalmClient", "FaceClient", "IrisClient" };
		//final String[] licenses = { "FingerFastExtractor", "PalmClient", "FaceFastExtractor", "IrisFastExtractor" };

		//=========================================================================

		//=========================================================================
		// TRIAL MODE
		//=========================================================================
		// Below code line determines whether TRIAL is enabled or not. To use purchased licenses, don't use below code line.
		// GetTrialModeFlag() method takes value from "Bin/Licenses/TrialFlag.txt" file. So to easily change mode for all our examples, modify that file.
		// Also you can just set TRUE to "TrialMode" property in code.
		//=========================================================================

		try {
			boolean trialMode = Utils.getTrialModeFlag();
			NLicenseManager.setTrialMode(trialMode);
			System.out.println("\tTrial mode: " + trialMode);
		} catch (IOException e) {
			e.printStackTrace();
		}

		NBiometricEngine engine = null;
		boolean anyMatchingComponent = false;
		try {
			// Obtain licenses
			for (String license : licenses) {
				if (NLicense.obtain("/local", 5000, license)) {
					System.out.format("Obtained license: %s%n", license);
					anyMatchingComponent = true;
				}
			}
			if (!anyMatchingComponent) {
				System.err.println("Could not obtain any matching license");
				System.exit(-1);
			}

			// Read CbeffRecord buffer
			NBuffer packedCBEFFRecord = NFile.readAllBytes(args[0]);

			// Get CbeffRecord patron format
			// all supported patron formats can be found in CbeffRecord class documentation
			int patronFormat = Integer.parseInt(args[1], 16);

			// Creating CbeffRecord object from NBuffer object
			CBEFFRecord cbeffRecord = new CBEFFRecord(packedCBEFFRecord, patronFormat);
			NSubject subject = new NSubject();
			engine = new NBiometricEngine();

			// Setting CbeffRecord
			subject.setTemplate(cbeffRecord);

			// Extracting template details from specified CbeffRecord data
			engine.createTemplate(subject);
			if (subject.getStatus() == NBiometricStatus.OK) {
				NFile.writeAllBytes(args[2], subject.getTemplate().save());
				System.out.println("Template successfully saved");
			} else {
				System.out.format("Template creation failed! Status: %s", subject.getStatus());
			}

		} catch (Throwable th) {
			Utils.handleError(th);
		} finally {
			if (engine != null) {
				engine.close();
			}
		}
	}
}

File: Tutorials/BiometricStandards/C/CbeffRecordToNTemplate/CbeffRecordToNTemplate.c
#include <TutorialUtils.h>

#ifdef N_APPLE_FRAMEWORKS
	#include <NCore/NCore.h>
	#include <NBiometrics/NBiometrics.h>
	#include <NMedia/NMedia.h>
	#include <NLicensing/NLicensing.h>
#else
	#include <NCore.h>
	#include <NBiometrics.h>
	#include <NMedia.h>
	#include <NLicensing.h>
#endif

const NChar title[] = N_T("CbeffRecordToNTemplate");
const NChar description[] = N_T("Converting CbeffRecord to NTemplate");
const NChar version[] = N_T("13.1.0.0");
const NChar copyright[] = N_T("Copyright (C) 2014-2024 Neurotechnology");

#define MAX_LICENSES 4

int usage()
{
	printf(N_T("usage:\n"));
	printf(N_T("\t%s [CbeffRecord] [PatronFormat] [NTemplate]\n"), title);
	printf(N_T("\n"));
	printf(N_T("\t[NTemplate] - filename of NTemplate\n"));
	printf(N_T("\t[CbeffRecord] - filename for CbeffRecord\n"));
	printf(N_T("\t[PatronFormat] - hex number identifying patron format (all supported values can be found in CbeffRecord class documentation)\n"));
	return 1;
}

int main(int argc, NChar **argv)
{
	NResult result = N_OK;
	HNBuffer hBuffer = NULL;
	HNBuffer hNTemplateBuffer = NULL;
	HCbeffRecord hCbeffRecord = NULL; // handle to CbeffRecord object
	HNSubject hSubject = NULL;
	HNBiometricEngine hBiometricEngine = NULL;
	HNString hBiometricStatus = NULL;
	const NChar * szBiometricStatus = NULL;
	NUInt patronFormat = 0;
	NBiometricStatus biometricStatus;
	NInt obtainedLicensesCount = 0, i = 0;
	
	//=========================================================================
	// CHOOSE LICENCES !!!
	//=========================================================================
	// ONE of the below listed "licenses" lines is required for unlocking this sample's functionality. Choose licenses that you currently have on your device.
	// If you are using a TRIAL version - choose any of them.

	const NChar * Licenses[MAX_LICENSES] = { N_T("FingerClient"), N_T("PalmClient"), N_T("FaceClient"), N_T("IrisClient") };
	//const NChar * Licenses[MAX_LICENSES] = { N_T("FingerFastExtractor"), N_T("PalmClient"), N_T("FaceFastExtractor"), N_T("IrisFastExtractor") };

	//=========================================================================

	OnStart(title, description, version, copyright, argc, argv);

	if(argc != 4)
	{
		OnExit();
		return usage();
	}

	//=========================================================================
	// TRIAL MODE
	//=========================================================================
	// Below code line determines whether TRIAL is enabled or not. To use purchased licenses, don't use below code line.
	// GetTrialModeFlag() method takes value from "Bin/Licenses/TrialFlag.txt" file. So to easily change mode for all our examples, modify that file.
	// Also you can just set TRUE to "TrialMode" property in code.
	{
		NBool isTrialMode = NFalse;
		result = GetTrialModeFlag(&isTrialMode);
		if (NFailed(result))
		{
			goto FINALLY;
		}
		result = NLicManSetTrialMode(isTrialMode);
		if (NFailed(result))
		{
			goto FINALLY;
		}
		printf(N_T("Trial mode: %d\n"), (int)isTrialMode);
	}
	//=========================================================================

	// check the license first
	for (i = 0; i < MAX_LICENSES; i++)
	{
		NBool available;
		result = NLicenseObtain(N_T("/local"), N_T("5000"), Licenses[i], &available);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("NLicenseObtain() failed, result = %d\n"), result);
			goto FINALLY;
		}
		if (available)
		{
			obtainedLicensesCount++;
		}
	}

	if (obtainedLicensesCount == 0)
	{
		printf(N_T("Could not obtain any matching license\n"));
		result = N_E_NOT_ACTIVATED;
		goto FINALLY;
	}

	// read CbeffRecord
	result = NFileReadAllBytesCN(argv[1], &hBuffer);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NFileReadAllBytesN() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// get CbeffRecord patron format
	result = NUInt32Parse(argv[2], N_T("X"), &patronFormat);
	if (NFailed(result))
	{ 
		result = PrintErrorMsgWithLastError(N_T("NUInt32Parse() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// create CbeffRecord from buffer
	result = CbeffRecordCreateFromMemoryNEx(hBuffer, patronFormat, 0, NULL, &hCbeffRecord);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("CbeffRecordCreateFromMemoryN() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// create subject for CbeffRecord
	result = NSubjectCreate(&hSubject);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NSubjectCreate() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// set CbeffRecord on subject
	result = NSubjectSetTemplateCbeff(hSubject, hCbeffRecord);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NSubjectSetTemplateCbeff() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// create BiometricEngine
	result = NBiometricEngineCreate(&hBiometricEngine);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NBiometricEngineCreate() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// create extract template information
	result = NBiometricEngineCreateTemplate(hBiometricEngine, hSubject, &biometricStatus);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NBiometricEngineCreateTemplate() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// check if template was sucessfully extracted
	if (biometricStatus != nbsOk)
	{
		// Retrieve biometric status
		result = NEnumToStringP(N_TYPE_OF(NBiometricStatus), biometricStatus, NULL, &hBiometricStatus);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("NEnumToStringP() failed (result = %d)!"), result);
			goto FINALLY;
		}

		result = NStringGetBuffer(hBiometricStatus, NULL, &szBiometricStatus);
		if (NFailed(result))
		{
			result = PrintErrorMsgWithLastError(N_T("NStringGetBuffer() failed (result = %d)!"), result);
			goto FINALLY;
		}

		printf(N_T("template extraction failed!\n"));
		printf(N_T("biometric status = %s.\n\n"), szBiometricStatus);

		result = N_E_FAILED;
		goto FINALLY;
	}

	// get NTemplate buffer
	result = NSubjectGetTemplateBuffer(hSubject, &hNTemplateBuffer);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NSubjectGetTemplateBuffer() failed, result = %d\n"), result);
		goto FINALLY;
	}

	// save CbeffRecord buffer to file
	result = NFileWriteAllBytesCN(argv[3], hNTemplateBuffer);
	if (NFailed(result))
	{
		result = PrintErrorMsgWithLastError(N_T("NFileWriteAllBytesCN() failed, result = %d\n"), result);
		goto FINALLY;
	}

	result = N_OK;
FINALLY:
	{
		NResult result2 = NObjectSet(NULL, &hBuffer);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hNTemplateBuffer);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hCbeffRecord);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hSubject);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NObjectSet(NULL, &hBiometricEngine);
		if NFailed((result2)) PrintErrorMsg(N_T("NObjectSet() failed (result = %d)!"), result2);
		result2 = NStringSet(NULL, &hBiometricStatus);
		if NFailed((result2)) PrintErrorMsg(N_T("NStringSet() failed (result = %d)!"), result2);
	}

	OnExit();
	return result;
}
